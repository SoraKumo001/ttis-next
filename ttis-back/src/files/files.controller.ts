import { Controller, Get, Param, Res } from '@nestjs/common';
import { FilesService } from './files.service';
import { FastifyReply } from 'fastify';
@Controller('files')
export class FilesController {
  constructor(private readonly service: FilesService) {}

  @Get(':id')
  async getTest(@Param('id') id: string, @Res() res: FastifyReply) {
    const { service } = this;
    const file = await service.getFile(id);
    if (!file) {
      res.status(404);
      res.raw.end('not found');
      return;
    }
    let httpDisposition = 'inline;';
    const fileName = file.name as string;
    const ext = fileName.split('.').pop();
    let contentType = 'application/octet-stream;';
    if (ext) {
      switch (ext.toLowerCase()) {
        case 'txt':
          contentType = 'text/plain';
          break;
        case 'html':
          contentType = 'text/html';
          break;
        case 'png':
          contentType = 'image/png';
          break;
        case 'svg':
          contentType = 'image/svg+xml';
          break;
        case 'jpeg':
        case 'jpg':
          contentType = 'image/jpeg';
          break;
        case 'gif':
          contentType = 'image/gif';
          break;
        default:
          httpDisposition = 'attachment;';
          break;
      }
    }
    res.header('Content-type', contentType);
    res.header('Content-length', (file.size as number).toString());
    res.header('Last-Modified', new Date(file.date).toUTCString());
    res.header(
      'Content-Disposition',
      `${httpDisposition} filename*=utf-8'jp'${encodeURI(fileName)}`,
    );
    res.raw.end(file.value, 'binary');
  }
}
