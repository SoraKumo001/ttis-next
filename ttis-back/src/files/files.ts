import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  Tree,
  TreeChildren,
  TreeParent,
  Index,
} from 'typeorm';
import { ObjectType, Field, ID, Int } from '@nestjs/graphql';

@ObjectType()
@Entity()
@Index(['parent', 'name'], { unique: true })
@Tree('materialized-path')
export class Files {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id!: string;
  @Field(() => Int)
  @Column({ default: 0 })
  kind!: number;
  @Field()
  @Column()
  name!: string;
  @Field({ nullable: true })
  @Column({ nullable: true })
  parentId?: string;
  @Field({ nullable: true })
  @TreeParent()
  parent?: Files;
  @Field(() => [Files], { nullable: true })
  @TreeChildren()
  children?: Files[];
  @Field(() => Int)
  size!: number;
  @Column({ type: 'bytea', nullable: true })
  value?: Buffer;
  @Field()
  @CreateDateColumn()
  createAt!: Date;
  @Field()
  @UpdateDateColumn()
  updateAt!: Date;
}
