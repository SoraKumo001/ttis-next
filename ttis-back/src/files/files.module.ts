import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Files } from './files';
import { FilesResolver } from './files.resolver';
import { FilesService } from './files.service';
import { FilesController } from './files.controller';

@Module({
  imports: [TypeOrmModule.forFeature([Files])],
  providers: [FilesService, FilesResolver],
  controllers: [FilesController],
})
export class FilesModule {}
