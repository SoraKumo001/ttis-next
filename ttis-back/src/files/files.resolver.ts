import { Resolver, Args, Info, ID, Int, Query, Mutation } from '@nestjs/graphql';
import { FilesService } from './files.service';
import { JwtAuthGuard, CurrentUser } from 'src/auth/auth.guard';
import { UseGuards } from '@nestjs/common';
import { User } from '@graphql/types';
import { GraphQLUpload, FileUpload } from 'graphql-upload';
import { createWriteStream } from 'fs';

import { Files } from './files';

@Resolver('Files')
export class FilesResolver {
  constructor(private readonly service: FilesService) {}

  @UseGuards(JwtAuthGuard)
  @Query(() => [Files], { nullable: true })
  async dirTree(@CurrentUser() user: User) {
    if (!user) return null;
    const { service } = this;
    return service.getDirList();
  }

  @UseGuards(JwtAuthGuard)
  @Mutation(() => Files, { nullable: true })
  async createDir(
    @CurrentUser() user: User,
    @Args('id', { type: () => ID }) id: string,
    @Args('name') name: string,
  ) {
    if (!user) return null;
    const { service } = this;
    return service.createDir(id,name);
  }
  @UseGuards(JwtAuthGuard)
  @Mutation(() => Files, { nullable: true })
  async renameFile(
    @CurrentUser() user: User,
    @Args('id', { type: () => ID }) id: string,
    @Args('name') name: string,
  ) {
    if (!user) return null;
    const { service } = this;
    return service.rename(id,name);
  }
  @UseGuards(JwtAuthGuard)
  @Mutation(() => Boolean, { nullable: true })
  async moveFile(
    @CurrentUser() user: User,
    @Args('targetId', { type: () => ID }) targetId: string,
    @Args('id', { type: () => ID }) id: string,
  ) {
    if (!user) return null;
    const { service } = this;
    return service.move(targetId,id);
  }

  @Mutation(() => Boolean, { nullable: true })
  async uploadFile(@Args({name: 'file', type: () => GraphQLUpload})
  {
      createReadStream,
      filename
  }: FileUpload): Promise<boolean> {
      return new Promise(async (resolve, reject) => 
          createReadStream()
              .pipe(createWriteStream(`./uploads/${filename}`))
              .on('finish', () => resolve(true))
              .on('error', () => reject(false))
      );
  }
}
