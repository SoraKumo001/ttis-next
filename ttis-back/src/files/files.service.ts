import { Injectable } from '@nestjs/common';
import { ExtendRepository } from 'src/libs/ExtendRepository';
import { Files } from './files';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';

export interface FileInfo {
  id: string;
  parentId?: string;
  kind: number;
  name: string;
  size: number;
  createAt: string;
  updateAt: string;
  childs?: FileInfo[];
}

@Injectable()
export class FilesService {
  constructor(
    @InjectRepository(Files)
    private readonly rep: ExtendRepository<Files>,
  ) {
    this.rep = new ExtendRepository(rep.manager.connection, Files);
  }
  async onModuleInit() {
    const { rep } = this;
   // await rep.count().then((count) => {
      // if (!count) {
      //   return rep.save({ kind: 0, name: "[ROOT]" });
      // }
    //});
  }
  public async getDirList() {
    const { rep } = this;

    const result = (await rep
      .createQueryBuilder()
      .select(`id,"parentId",kind,name,"createAt","updateAt",octet_length(value) as size`)
      .where("kind=0")
      .orderBy("name")
      .getRawMany()) as FileInfo[];
    // const hash = new Map<string, FileInfo>();
    // if (dirInfos) {
    //   for (const dir of dirInfos) {
    //     dir.childs = [];
    //     hash.set(dir.id, dir);
    //   }
    //   for (const dir of hash.values()) {
    //     const parentId = dir.parentId;
    //     if (parentId) {
    //       const p = hash.get(parentId);
    //       if (p && p.childs) p.childs.push(dir);
    //     }
    //   }
    // }

    return result;
  }
}
