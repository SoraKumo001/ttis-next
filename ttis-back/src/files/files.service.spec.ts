import { Test, TestingModule } from '@nestjs/testing';
import { FilesService } from './files.service';
import { DBModule2 } from 'src/db.module';
import { FilesModule } from './files.module';
import { ObjectMask } from '@test/testTools';

const masks = {
  id: '',
  parentId: '',
  createAt: '',
  updateAt: '',
};

describe('FileService', () => {
  let service: FilesService;
  let module: TestingModule;
  beforeAll(async () => {
    module = await Test.createTestingModule({
      imports: [DBModule2, FilesModule],
    }).compile();
    await module.init();

    service = module.get<FilesService>(FilesService);
  });
  afterAll(async () => {
    await module.close();
  });
  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('Create dir', async () => {
    const dir = await service.createDir(undefined,"test");
    expect(dir).toBeDefined();
    const dir2 = await service.createDir(undefined,"test");
    expect(dir2).toEqual(null);
  });
  it('List Contents first', async () => {
    const list = await service.getDirList();
    expect(ObjectMask(list, masks)).toMatchSnapshot();
  });
});
