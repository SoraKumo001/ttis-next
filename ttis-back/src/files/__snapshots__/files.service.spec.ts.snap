// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileService List Contents first 1`] = `
Array [
  Object {
    "createAt": 2020-06-04T17:47:24.903Z,
    "id": "fa2354cd-934e-458b-bd87-6bcaaee50ae3",
    "kind": 0,
    "name": "[ROOT]",
    "parentId": null,
    "size": null,
    "updateAt": 2020-06-04T17:47:24.903Z,
  },
  Object {
    "createAt": 2020-06-04T17:47:24.926Z,
    "id": "4d8ab2b5-24f3-4d77-aa8e-fbb23e8c14a6",
    "kind": 0,
    "name": "test",
    "parentId": "fa2354cd-934e-458b-bd87-6bcaaee50ae3",
    "size": null,
    "updateAt": 2020-06-04T17:47:24.926Z,
  },
]
`;
